### ISSUE TITLE
Inverted `Promise` should warn like it does without inverting

### ISSUE THREAD

**Issue by soffes** [microsoft/TypeScript#46140]
# Suggestion

## 🔍 Search Terms

> This condition will always return true since this 'Promise<string | undefined>' is always defined.(2801) Did you forget to use 'await'?

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

When checking the inverted truthyness if something returned from an `async` function without calling `await`, it should warn you you forgot `await`. This works without inverting.

## 📃 Motivating Example

```ts
async function foo(): Promise<string | undefined> {
    return
}

const value = foo()
if (!value) { // <-- fails silently
    // ...
}

const awaitedValue = await foo()
if (awaitedValue) { // <-- Emits ts2801
    // ...
}
```

[Playground](https://www.typescriptlang.org/play?#code/IYZwngdgxgBAZgV2gFwJYHsL3egFASgC4YAFAJ3QFtUQBTAHhGTNQgHMYAfGJAE1ritavAHwwA3gCgYMmGVrIEZCJIC+kyVExMYAN2AAbBLRgBebHnyTUcGLgCE+o7XwTpsgPQeYAOj9rJIA)

## 💻 Use Cases

In my use, I had a function that returned a `User | undefined` and I was doing something if there wasn't a user returned. Later, I made the function async and forgot to update this since it was failing silently.

**Comment by andrewbranch** [comment]
This should probably be predicated on the possibility that the awaited type of the Promise is possibly falsy. I think we could look at this for uncalled functions too, since that shares most of the same code path.

**Comment by IllusionMH** [comment]
First case looks similar to #44840

**Comment by andrewbranch** [comment]
Yeah, there are several issues asking us to be more aggressive about either recognizing code as unreachable or issuing errors when we already do (like here—`value` is successfully narrowed to `never` inside `if (!value)`, but the user has no way of noticing that because they think they just detected it to be `undefined`, so of course they’re not going to try to use it). But I like this issue because the scope is very narrow, and seems to sit well within the unawaited Promise / uncalled function checks we already have.

Also related is #45267

**Comment by fatcerberus** [comment]
> ...but the user has no way of noticing that because they think they just detected it to be `undefined`, so of course they’re not going to try to use it

One potential issue: narrowing to `never` is not guaranteed to produce an error even if the value is used, since `never` is assignable to all other types.

**Comment by Miodec** [comment]
Any update? Just ran into this issue. Tried 4.6.4 and its still a problem.

The bug I found today:
```ts

export async function updateName(
  uid: string,
  name: string
): Promise<UpdateResult> {
  if (!isNameAvailable(name)) {
    throw new MonkeyError(409, "Username already taken", name);
  }

  const user = await getUser(uid, "update name");
...
```

`isNameAvailable` returns a promise, and so the throw line will never execute.

Oddly enough, removing the inversion highlights this error correctly:
```ts
if (isNameAvailable(name)) {
```
```
This condition will always return true since this 'Promise<boolean>' is always defined.ts(2801)
user.ts(54, 7): Did you forget to use 'await'?
```

ANOTHER interesting fact is that using `===` also highlights this error correctly:
```ts
if (!isNameAvailable(name) === true) {
```
```
This condition will always return 'false' since the types 'false' and 'true' have no overlap.ts(2367)
```

But, as mentioned, the first code block is silently always false, causing headaches 😄

**Comment by tjenkinson** [comment]
Hey! We had a bug that would also have been caught by this.

I started trying to implement a solution here: https://github.com/microsoft/TypeScript/compare/main...tjenkinson:TypeScript:support-condition-always-true-in-prefix-unary-expression

Let me know if it looks like the right approach and I'll make it a PR and add some tests :)

**Comment by gabritto** [comment]
> Hey! We had a bug that would also have been caught by this.
> 
> I started trying to implement a solution here: [main...tjenkinson:TypeScript:support-condition-always-true-in-prefix-unary-expression](https://github.com/microsoft/TypeScript/compare/main...tjenkinson:TypeScript:support-condition-always-true-in-prefix-unary-expression)
> 
> Let me know if it looks like the right approach and I'll make it a PR and add some tests :)

It looks ok at a first glance.
I noticed one thing missing.
When the thing that is always truthy has a function type, right now we only error if the thing is not used in the body of the condition. Example:
```ts
{
    let fun!: () => void;

    if (fun) { // Error
        let x = 1;
    }
}

{
    let fun!: () => void;

    if (fun) { // No error, because `fun` could actually be undefined and this is a valid check
        let x = fun();
    }
}
```
For the negative case, the equivalent would be this:
```ts
{
    let fun!: () => void;

    if (!fun) { // Should be error, probably a bug
        // ...
    }
}

{
    let fun!: () => void;

    if (!fun) { // Should not be error, `fun` could actually be undefined and this check is valid
        // ...
    }
    else {
        let x = fun();
    }
}
```
so your implementation would have to account for that.

It would also be good to have more tests, for the other possible error and non-error cases for instance (e.g. unawaited promises), if you want to move forward with this.

**Comment by tjenkinson** [comment]
thanks @gabritto 

I'm wondering what the reasoning is behind ignoring when the variable is referenced in the block? Is it because there are a lot of cases in the wild where the types are wrong and therefore logic is guarded in if statements just in case?

If that's the case I wonder if the same really applies when checking the inverse?

If it does I guess we'd also need to allow code like

```ts
let fun!: () => void;

if (!fun) {
    throw new Error('`fun` missing');
}

fun();
```
which looks more complicated given the call is not part of the `if/else`?

Wondering if when using `!` we should only check the promise case?

I.e.

```ts
let isEnabled!: () => Promise<boolean>;

// always error
if (!isEnabled()) {}

// always ok
if (!isEnabled) {}
```

WDYT?

**Comment by gabritto** [comment]
> I'm wondering what the reasoning is behind ignoring when the variable is referenced in the block? Is it because there are a lot of cases in the wild where the types are wrong and therefore logic is guarded in if statements just in case?

Yes.

> If that's the case I wonder if the same really applies when checking the inverse?

We can start looking into that by creating a PR with the original version of your code (not doing anything special for promises or the negation case), and then I can run the extended tests on it to see if we need any special casing or not.

**Comment by tjenkinson** [comment]
@gabritto sounds great opened https://github.com/microsoft/TypeScript/pull/60068 and it's back to checking everything
